src/DGtal/topology/CubicalComplex.h

```diff
@@ -0,0 +1,1307 @@
+          {
+            if ( ++myD > myCC->dimension ) break;
+            myIt    = myCC->begin( myD );
+            myItEnd = myCC->end( myD );
+          }
+      }
+
+      void increment()
+      {
+        ASSERT( myCC != 0 );
+        ++myIt;
+        nextDimension();
+      }
+
+      bool equal( const Iterator& other ) const
+      { 
+        return ( myD == other.myD ) && ( myIt == other.myIt );
+      }
+
+      Cell const& dereference() const
+      { 
+        return myIt->first;
+      }
+
+      Dimension dimension() const
+      { 
+        return myD;
+      }
+      
+    private:
+      CubicalComplex* myCC;
+      Dimension myD;
+      CellMapIterator myIt;
+      CellMapIterator myItEnd;
+    };
+
+    // ----------------------- STL inner types ------------------------------
+  public:
+
+    // Renaming for STL-type of iterator.
+    typedef ConstIterator                           const_iterator;
+    typedef Iterator                                iterator;
+    typedef Cell                                    value_type;
+    typedef Cell const&                             reference;
+    typedef Cell const&                             const_reference;
+    typedef typename CellContainer::size_type       size_type; 
+    typedef typename CellContainer::difference_type difference_type;
+    typedef Cell const*                             pointer;
+    typedef Cell const*                             const_pointer;
+
+    // ----------------------- Standard services ------------------------------
+  public:
+
+    /**
+    * Destructor.
+    */
+    ~CubicalComplex();
+
+  protected:
+    /**
+    * Constructor. The cubical complex is not valid. A user may not
+    * instantiate an empty CubicalComplex, because it needs the
+    * Khalimsky space.
+    */
+    CubicalComplex();
+  public:
+
+    /**
+    * Constructor of empty complex. Needs a space to represents
+    * cubical cells.
+    *
+    * @param aK a Khalimsky space.
+    */
+    CubicalComplex ( ConstAlias<KSpace> aK );
+
+    /**
+    * Copy constructor.
+    * @param other the object to clone.
+    * Forbidden by default.
+    */
+    CubicalComplex ( const CubicalComplex & other );
+    
+    /**
+    * Constructor a complex from a digital set.
+    * @param set - a digital set from which to create a complex. 
+    * Set has to be of the same dimension as a Khalimsky space.
+    */
+    template < typename TDigitalSet >
+    void construct ( const TDigitalSet & set );
+
+    /**
+    * Assignment.
+    * @param other the object to copy.
+    * @return a reference on 'this'.
+    * Forbidden by default.
+    */
+    CubicalComplex & operator= ( const CubicalComplex & other );
+
+    /**
+    * Clears the cubical complex, which becomes empty.
+    */
+    void clear();
+
+    /**
+    * Clears all cells of dimension \a d of the cubical complex.
+    * @param d the dimension of cell \a aCell.
+    */
+    void clear( Dimension d );
```

Comment:
It is passed by value, so there is no difference for the user, and more time consuming for him to add a const. Some people do advise to add the const only in the definition of the method (not the declaration). Even so, it is so easy for a compiler to detect that a variable is only read that I do not see what prevents him from optimizing it.